consensus/reactor.go

```diff
@@ -877,28 +877,43 @@ func (conR *Reactor) peerStatsRoutine() {
 		}
 
 		select {
-		case msg := <-conR.conS.statsMsgQueue:
-			// Get peer
-			peer := conR.Switch.Peers().Get(msg.PeerID)
-			if peer == nil {
-				conR.Logger.Debug("Attempt to update stats for non-existent peer",
-					"peer", msg.PeerID)
-				continue
-			}
-			// Get peer state
-			ps, ok := peer.Get(types.PeerStateKey).(*PeerState)
-			if !ok {
-				panic(fmt.Sprintf("Peer %v has no state", peer))
-			}
-			switch msg.Msg.(type) {
-			case *VoteMessage:
-				if numVotes := ps.RecordVote(); numVotes%votesToContributeToBecomeGoodPeer == 0 {
-					conR.Switch.MarkPeerAsGood(peer)
+		case mInterface := <-conR.conS.peerInfoQueue:
+			switch msg := mInterface.(type) {
+			case msgInfo:
+				// Get peer
+				peer := conR.Switch.Peers().Get(msg.PeerID)
+				if peer == nil {
+					conR.Logger.Debug("Attempt to update stats for non-existent peer",
+						"peer", msg.PeerID)
+					continue
 				}
-			case *BlockPartMessage:
-				if numParts := ps.RecordBlockPart(); numParts%blocksToContributeToBecomeGoodPeer == 0 {
-					conR.Switch.MarkPeerAsGood(peer)
+				// Get peer state
+				ps, ok := peer.Get(types.PeerStateKey).(*PeerState)
+				if !ok {
+					panic(fmt.Sprintf("Peer %v has no state", peer))
+				}
+				switch msg.Msg.(type) {
+				case *VoteMessage:
+					if numVotes := ps.RecordVote(); numVotes%votesToContributeToBecomeGoodPeer == 0 {
+						conR.Switch.MarkPeerAsGood(peer)
+					}
+				case *BlockPartMessage:
+					if numParts := ps.RecordBlockPart(); numParts%blocksToContributeToBecomeGoodPeer == 0 {
+						conR.Switch.MarkPeerAsGood(peer)
+					}
+				}
+
+			case badPeerInfo:
+				err, peerID := msg.Error, msg.PeerID
+				peer := conR.Switch.Peers().Get(peerID)
+				if peer == nil {
+					conR.Logger.Debug("Attempt to remove non-existent peer",
+						"peer", peerID)
+					continue
 				}
```

Comment:
If Get can fail, it needs to return an error, failure should not be expressed as a nil peer.

Reply:
Its not a failure, its just a miss - there's no value under that key.

Reply:
A read miss is a failure, you can't say it's equivalent to the default value of the value type of the collection.

```go
m := map[int]int{31: 0}
v0 := m[0]  // read miss resulting in v0 == 0
v31 := m[31] // read hit resulting in v31 == 0
```

Get has to return `(ValueType, error)` or at least `(ValueType, bool)`. No way around it.

Reply:
@peterbourgon in your example, `v12` can validly be `0`, but a peer can't validly be `nil` here.

While I understand that it's more idiomatic, it seems like addressing that would add complexity for very little value (we'd need to add a specific error type for when the peer doesn't exist, and then check that error's type to ensure that it was a miss and not some other error, and then what do we do with other errors?).